lib/cchain/provider/abci.go

```diff
@@ -357,40 +356,67 @@ func newABCIFetchFunc(attCl atypes.QueryClient, cmtCl cmtservice.ServiceClient,
 		atts, ok, err := attsFromAtHeight(ctx, attCl, chainVer, fromOffset, 0)
 		if err != nil {
 			incQueryErr(endpoint)
-			return nil, errors.Wrap(err, "abci query attestations-from")
+			return nil, 0, errors.Wrap(err, "abci query attestations-from")
 		} else if ok {
-			fetchStepsMetrics(chainName, 0, 0)
-			return atts, nil
+			binarySearchStepsMetric(chainName, 0)
+			lookbackStepsMetric(chainName, 0)
+
+			return atts, 0, nil
 		}
 
 		earliestAttestationAtLatestHeight, ok, err := queryEarliestAttestation(ctx, attCl, chainVer, 0)
 		if err != nil {
 			incQueryErr(endpoint)
-			return nil, errors.Wrap(err, "abci query earliest-attestation-in-state")
+
+			return nil, 0, errors.Wrap(err, "abci query earliest-attestation-in-state")
 		}
 
 		// Either no attestations have happened yet, or the queried fromOffset is in the "future"
 		// Caller has to wait and retry in both cases
 		if !ok || earliestAttestationAtLatestHeight.AttestOffset < fromOffset {
 			// First attestation hasn't happened yet, return empty
-			return []xchain.Attestation{}, nil
+			return []xchain.Attestation{}, 0, nil
 		}
 
-		offsetHeight, err := searchOffsetInHistory(ctx, cmtCl, attCl, chainVer, chainName, fromOffset)
+		latestBlockResp, err := cmtCl.GetLatestBlock(ctx, &cmtservice.GetLatestBlockRequest{})
+		if err != nil {
+			return []xchain.Attestation{}, 0, errors.Wrap(err, "query latest block")
+		}
+		latestHeight := uint64(latestBlockResp.SdkBlock.Header.Height)
+
+		var startHeight, endHeight uint64
+		// if we are performing search for the first time and no previous height was cached do a full search
+		if cachedHeight == 0 {
+			startHeight, endHeight, err = lookbackRange(ctx, attCl, chainVer, chainName, fromOffset, latestHeight)
+			if err != nil {
+				incQueryErr(endpoint)
+
+				return nil, 0, errors.Wrap(err, "lookback search")
+			}
+		} else {
+			// otherwise use cached height from previous search as the next start height
+			// but increased by one since we are searching the next attestation offset
+			startHeight = cachedHeight + 1
```

Comment:
as mentioned before, there is no guarantee that this offset isn't the same height as previous fetch. You can't assume this. Remove the +1.